### ISSUE TITLE
RepetableRead isolation level leads to concurrency issues

### ISSUE THREAD

**Issue by Frazi1** [frankhommers/Hangfire.PostgreSql#191]
## Synopsis

At my work, we use Hangfire.Postgres + [Hangfire.Tags](https://github.com/face-it/Hangfire.Tags) libraries in production. We use it mostly for scheduling. In total, we have ~40 recurring jobs. Each job, when executed, just enqueues other normal jobs to Hangfire. Having that, the most common case when we experience concurrency issues is __enqueing__  a job. The error we are getting is:

```c#
Hangfire.BackgroundJobClientException: Background job creation failed. See inner exception for details.
 ---> Npgsql.PostgresException (0x80004005): 40001: could not serialize access due to concurrent update
```

Originally I opened an issue in Hangfire.Tags repository and provided a fix in there. However, @erwin-faceit (author of Hangfire.Tags) had a very good point that such concurrency issues need to be addressed at the storage level because any code that uses `PostgreSqlWriteOnlyTransaction` can be victim to concurrency issue in high-load scenarios. So, it affects Hangfire.Postgres and any other extension that is built on top of Hangfire.Postgres.

See our conversation in Hangfire.Tags repo for more details: https://github.com/face-it/Hangfire.Tags/issues/35#issuecomment-824781655

## Postgres, transactions, and concurrency issues
Hangfire.Postgres uses `RepeatableRead` isolation level for its transactions. See `Commit` method here: https://github.com/frankhommers/Hangfire.PostgreSql/blob/11eb27ae5f5d5817876cafe5cf11760d0853277c/src/Hangfire.PostgreSql/PostgreSqlWriteOnlyTransaction.cs#L55-L87

Postgres [documentation](https://www.postgresql.org/docs/9.1/transaction-iso.html)(1.3.2.2. Repeatable Read Isolation Level section) clearly says the following about `RepeatableRead` isolation level:

> Applications using this level must be prepared to retry transactions due to serialization failures.

However, Hangfire.Postgres does not implement any retry logic.

## Reproduction steps

I've built a repo to reproduce this issue. It requires Hangfire.Tags package to be installed.

Please see my comment https://github.com/face-it/Hangfire.Tags/issues/35#issuecomment-791813078. It contains all the information: 
* links
* step-by-step repro guide
* stack traces
* screenshots.

## Possible solution

We have 2 alternative ways to deal with this problem:
1. Keep `RepeatableRead` isolation level. Implement logic to retry transactions that failed due to concurrency issue
    * See example of a retry-policy here: https://github.com/face-it/Hangfire.Tags/blob/310a9b2a77f7ef331f99d55f79b9cf83495a0739/src/Hangfire.Tags.PostgreSql/PostgreSqlTransactionRetryPolicy.cs#L30-L56
2. Switch to using `ReadCommitted` isolation level.

Other storages use `ReadCommitted` isolation level. Please see https://github.com/face-it/Hangfire.Tags/issues/35#issuecomment-824781655 and https://github.com/face-it/Hangfire.Tags/issues/35#issuecomment-824784060 comments from @erwin-faceit for more details on this matter. Also, see Hangfire.SqlServer implementation: https://github.com/HangfireIO/Hangfire/blob/f9497dbf6ecc9202454be7b176681967d73e61ff/src/Hangfire.SqlServer/SqlServerStorage.cs#L271-L280.

I personally find retry-policy a safer solution, but switching the isolation might be the right way to go because it should provide a more reliable solution. With the retry policy, it is still not clear how many retries we want to have because it all depends on the number of competing workers. The logic can fail with concurrency errors dozens of times before it finally succeeds.

## Questions
Before we try to build any pull requests, I would like to get some ideas from other people.

I can see that `RepeatableRead` isolation level was implemented back in 2014 and has not been changed since then. Is there any particular reason why you choose it over `ReadCommitted` @frankhommers? Do you see any problem if we tried to make that change today?

@vytautask @yongliu-mdsol or anybody else - please comment if you have any thoughts or ideas. I would love to hear them.

__P.S.__ Sorry for the trouble having to follow links to other repos. I tried my best to document it, but I didn't want to copy-paste everything from one repo to another. Also, I wanted to give credit to @erwin-faceit because he was the one who suggested addressing this issue at the storage level.

**Comment by vytautask** [comment]
That's a very fine description of a problem, @Frazi1!
Looking back into previous commit history I came across the change made a while ago to ReadCommitted in JobQueue class (https://github.com/frankhommers/Hangfire.PostgreSql/commit/9b1413544981e2e032681866f3faec353befb946) - seems to be holding quite alright over all those years.
I would say we should switch that WriteOnlyTransaction to ReadCommitted as well.

**Comment by Frazi1** [comment]
@vytautask thanks for your comment!

I can try switching to `ReadCommitted` and verify that it resolves the issue. If it looks good, I'll prepare a PR.

Also, we can expose a setting for consumers to pick the isolation level. Do you think it can be a good safety measure to allow the users to quickly rollback to the older implementation?

**Comment by vytautask** [comment]
If we will get more issues, I guess we could just rollback.
New release will only contain this fix and could be rolled back on user side as well. So I would say an additional option is not needed.

**Comment by Frazi1** [comment]
> If we will get more issues, I guess we could just rollback. New release will only contain this fix and could be rolled back on user side as well. So I would say an additional option is not needed.

Ok, sounds good. I'll try to get a PR ready later today or tomorrow.

**Comment by vytautask** [comment]
Nice investigation!
Yup, it seems merge is a way better option than current implementation.
Also it supports really old postgresql versions as well.

**Comment by vytautask** [comment]
Should be fixed with 1.5.8.2 release (available in NuGet now).
